File: twister-lib-android/capture_image/src/main/java/net/twisterrob/android/view/ExternalPicker.java
package net.twisterrob.android.view;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.activity.ComponentActivity;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultCaller;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.PickVisualMediaRequest;
import androidx.activity.result.PickVisualMediaRequestKt;
import androidx.activity.result.contract.ActivityResultContract;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.graphics.drawable.DrawableWrapperCompat;
import androidx.appcompat.widget.PopupMenu;

import net.twisterrob.android.capture_image.R;
import net.twisterrob.android.content.ImageRequest;
import net.twisterrob.android.utils.tools.ViewTools;

/**
 * Abstraction for handling for picking from external image source.
 * Currently implemented via a {@link PopupMenu}.
 * <p> 
 * Note: "pick" is an overloaded term.
 * <ul>
 *     <li>"External Picker" and "pick" in CaptureImage activity means to import from external image sources.</li>
 *     <li>PickVisualMedia / MediaStore.ACTION_PICK_IMAGES is a new way of selecting images via the Photo Picker.</li>
 *     <li>Intent.ACTION_PICK is an old API for selecting anything from any other app.</li>
 * </ul>
 */
public class ExternalPicker {

	private final @NonNull Context context;
	private final @NonNull Events events;
	private final @NonNull PopupMenu menu;

	public ExternalPicker(
			@NonNull ComponentActivity activity,
			@NonNull View anchor,
			@NonNull Uri target,
			@NonNull Events events
	) {
		this.context = activity;
		this.events = events;
		ExplicitAbleActivityResultLauncher<PickVisualMediaRequest> pickImage = new ExplicitAbleActivityResultLauncher<>(
				activity,
				new ActivityResultContracts.PickVisualMedia(), // STOPSHIP real pick?
				PickVisualMediaRequestKt.PickVisualMediaRequest(
						ActivityResultContracts.PickVisualMedia.ImageOnly.INSTANCE
				),
				(@Nullable Uri result) -> {
					if (result != null) {
						events.onPick(result);
					} else {
						events.onCancelled();
					}
				}
		);
		ExplicitAbleActivityResultLauncher<String> getContent = new ExplicitAbleActivityResultLauncher<>(
				activity,
				new ActivityResultContracts.GetContent(),
				"image/*",
				(@Nullable Uri result) -> {
					if (result != null) {
						events.onGetContent(result);
					} else {
						events.onCancelled();
					}
				}
		);
		ExplicitAbleActivityResultLauncher<Uri> captureImage = new ExplicitAbleActivityResultLauncher<>(
				activity,
				new ActivityResultContracts.TakePicture(),
				target,
				(@Nullable Boolean result) -> {
					if (Boolean.TRUE.equals(result)) {
						events.onCapture(target);
					} else {
						events.onCancelled();
					}
				}
		);
		menu = createMenu(activity, anchor);
		menu.getMenu().findItem(R.id.image__choose_external__get).setIntent(getContent.createIntent());
		menu.getMenu().findItem(R.id.image__choose_external__pick).setIntent(pickImage.createIntent());
		menu.getMenu().findItem(R.id.image__choose_external__capture).setIntent(captureImage.createIntent());
		menu.setOnMenuItemClickListener((@NonNull MenuItem item) -> {
			// Clear the dismiss listener, so it doesn't get called after this method returns.
			// We only want to call it when the user cancels the menu, not when they select something.
			menu.setOnDismissListener(null);
			events.itemSelected();
			if (item.getItemId() == R.id.image__choose_external__get
					|| item.getGroupId() == R.id.image__choose_external__get_group) {
				getContent.launch(item.getIntent());
				return true;
			} else if (item.getItemId() == R.id.image__choose_external__pick
					|| item.getGroupId() == R.id.image__choose_external__pick_group) {
				pickImage.launch(item.getIntent());
				return true;
			} else if (item.getItemId() == R.id.image__choose_external__capture
					|| item.getGroupId() == R.id.image__choose_external__capture_group) {
				captureImage.launch(item.getIntent());
				return true;
			} else {
				throw new IllegalArgumentException("Unknown menu item: " + item);
			}
			// Execution continues in registerForActivityResult callbacks.
		});
	}

	private static @NonNull PopupMenu createMenu(@NonNull Activity activity, @NonNull View anchor) {
		PopupMenu menu = new PopupMenu(activity, anchor);
		menu.setGravity(Gravity.TOP | Gravity.START);
		menu.setForceShowIcon(true); // See fixIcons().
		menu.inflate(R.menu.image__choose_external);
		return menu;
	}

	public void show() {
		// Set dismiss listener right before showing to make sure it exists.
		// It might be cleared when an item is selected. See #setOnMenuItemClickListener.
		menu.setOnDismissListener(menu -> events.onCancelled());
		resolveIntents(menu.getMenu());
		menu.show();
	}

	private void resolveIntents(@NonNull Menu menu) {
		populate(menu, R.id.image__choose_external__pick, R.id.image__choose_external__pick_group);
		populate(menu, R.id.image__choose_external__get, R.id.image__choose_external__get_group);
		populate(menu, R.id.image__choose_external__capture, R.id.image__choose_external__capture_group);
		fixIcons(menu);
		showCapture(menu, ImageRequest.canLaunchCameraIntent(context));
	}

	private void populate(@NonNull Menu menu, @IdRes int itemId, @IdRes int groupId) {
		MenuItem item = menu.findItem(itemId);
		menu.addIntentOptions(groupId, Menu.NONE, item.getOrder() + 1, null, null, item.getIntent(), 0, null);
	}

	private void showCapture(@NonNull Menu menu, boolean canCapture) {
		MenuItem captureItem = menu.findItem(R.id.image__choose_external__capture);
		ViewTools.visibleIf(captureItem, canCapture);
		menu.setGroupVisible(R.id.image__choose_external__capture_group, captureItem.isVisible());
	}

	private void fixIcons(@NonNull Menu menu) {
		for (int i = 0; i < menu.size(); i++) {
			MenuItem item = menu.getItem(i);
			item.setIcon(fix(item.getIcon(), item.getGroupId() != Menu.NONE, context.getResources()));
		}
	}

	/**
	 * Ensure all drawables have the same size, otherwise they're misaligned.
	 * Also add some indentation to resolved package listings, to give some hierarchy without use of submenus.
	 * <p>
	 * {@link androidx.appcompat.view.menu.StandardMenuPopup}
	 * inflates {@link androidx.appcompat.R.layout.abc_popup_menu_item_layout}
	 * using {@link androidx.appcompat.view.menu.MenuAdapter}.
	 * Items in the adapter will be {@link androidx.appcompat.view.menu.ListMenuItemView}
	 * and their icons are added in {@link androidx.appcompat.view.menu.ListMenuItemView#insertIconView}
	 * from {@link androidx.appcompat.R.layout.abc_list_menu_item_icon}.
	 * @noinspection JavadocReference
	 */
	private static @Nullable Drawable fix(@Nullable Drawable icon, boolean sub, @NonNull Resources resources) {
		if (icon == null) {
			return null;
		}
		int indent = sub ? resources.getDimensionPixelSize(R.dimen.image__choose_external__menu_icon_indent) : 0;
		int size = resources.getDimensionPixelSize(R.dimen.image__choose_external__menu_icon_size);
		return new DrawableWrapperCompat(icon) {
			@Override public int getIntrinsicWidth() {
				// Adding indent reserves extra width, because ImageView will take this size.
				return size + indent;
			}
			@Override public int getIntrinsicHeight() {
				return size;
			}
			@Override public void setBounds(int left, int top, int right, int bottom) {
				// Shift left by `indent`, so the icon gets back to it's `size` width, and the left side of it will be empty.
				super.setBounds(left + indent, top, right, bottom);
			}
		};
	}

	public interface Events {
		void onCancelled();
		void itemSelected();
		void onGetContent(@NonNull Uri result);
		void onPick(@NonNull Uri result);
		void onCapture(@NonNull Uri result);
	}

	private static class ExplicitAbleActivityResultLauncher<I> {
		private final @NonNull Context context;
		private final @NonNull ActivityResultContract<I, ?> contract;
		private final @NonNull I input;
		private final @NonNull ActivityResultLauncher<Intent> launcher;

		private <O, C extends Context & ActivityResultCaller> ExplicitAbleActivityResultLauncher(
				@NonNull C context,
				@NonNull ActivityResultContract<I, O> contract,
				@NonNull I input, @NonNull ActivityResultCallback<O> callback
		) {
			this.context = context;
			this.contract = contract;
			this.input = input;
			this.launcher = context.registerForActivityResult(
					new ActivityResultContracts.StartActivityForResult(),
					ar -> callback.onActivityResult(
							contract.parseResult(ar.getResultCode(), ar.getData())
					)
			);
		}

		/**
		 * Create a generic intent for the {@link #contract} and {@link #input}.
		 */
		public @NonNull Intent createIntent() {
			return contract.createIntent(context, input);
		}

		/**
		 * Launch the generic intent generated by {@link #createIntent()},
		 * or launch a package-specific explicit intent (overridden component) from the same.
		 */
		public void launch(@NonNull Intent input) {
			launcher.launch(input);
		}
	}
}

File: twister-lib-android/capture_image/src/main/java/net/twisterrob/android/view/ExternalPicker.kt
package net.twisterrob.android.view

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.ActivityResultCaller
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContract
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.IdRes
import androidx.annotation.RequiresApi
import androidx.appcompat.graphics.drawable.DrawableWrapperCompat
import androidx.appcompat.widget.PopupMenu
import net.twisterrob.android.activity.CaptureImage
import net.twisterrob.android.capture_image.R
import net.twisterrob.android.contracts.PickForMimeType
import net.twisterrob.android.utils.tools.CameraTools
import net.twisterrob.android.utils.tools.TextTools
import net.twisterrob.android.utils.tools.asIterable

/**
 * Abstraction for handling for picking from external image source.
 * Currently implemented via a [PopupMenu].
 *
 * Note: "pick" is an overloaded term.
 *  * [ExternalPicker] and [CaptureImage.pick] means to import from external image sources.
 *  * [ActivityResultContracts.PickVisualMedia] / [MediaStore.ACTION_PICK_IMAGES] is a new way
 *    of selecting images via the Photo Picker.
 *  * [Intent.ACTION_PICK] is an old API for selecting anything from any other app.
 */
class ExternalPicker(
	activity: ComponentActivity,
	anchor: View,
	target: Uri,
	private val events: Events
) {
	private val context: Context
	private val menu: PopupMenu

	private val pickImage = ExplicitAbleActivityResultLauncher(
		activity,
		PickForMimeType(),
		"image/*",
		ActivityResultCallback { result: Uri? ->
			if (result != null) {
				events.onPick(result)
			} else {
				events.onCancelled()
			}
		}
	)

	private val getContent = ExplicitAbleActivityResultLauncher(
		activity,
		ActivityResultContracts.GetContent(),
		"image/*"
	) { result: Uri? ->
		if (result != null) {
			events.onGetContent(result)
		} else {
			events.onCancelled()
		}
	}

	@RequiresApi(Build.VERSION_CODES.KITKAT) // see isPhotoPickerAvailable
	private val pickVisualImage = ExplicitAbleActivityResultLauncher(
		activity,
		ActivityResultContracts.PickVisualMedia(),
		PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly)
	) { result: Uri? ->
		if (result != null) {
			events.onPickVisualImage(result)
		} else {
			events.onCancelled()
		}
	}

	private val captureImage = ExplicitAbleActivityResultLauncher(
		activity,
		ActivityResultContracts.TakePicture(),
		target
	) { result: Boolean? ->
		if (java.lang.Boolean.TRUE == result) {
			events.onCapture(target)
		} else {
			events.onCancelled()
		}
	}

	init {
		context = activity
		menu = createMenu(activity, anchor)
		menu.menu.apply {
			findItem(R.id.image__choose_external__pick).intent = pickImage.createIntent()
			findItem(R.id.image__choose_external__get).intent = getContent.createIntent()
			findItem(R.id.image__choose_external__visual).intent = pickVisualImage.createIntent()
			findItem(R.id.image__choose_external__capture).intent = captureImage.createIntent()
		}
		menu.setOnMenuItemClickListener { item: MenuItem ->
			// Clear the dismiss listener, so it doesn't get called after this method returns.
			// We only want to call it when the user cancels the menu, not when they select something.
			menu.setOnDismissListener(null)
			events.itemSelected()
			when {
				item.itemId == R.id.image__choose_external__get
					|| item.groupId == R.id.image__choose_external__get_group -> {
					getContent.launch(item.intent!!)
					true
				}
				item.itemId == R.id.image__choose_external__pick
					|| item.groupId == R.id.image__choose_external__pick_group -> {
					pickImage.launch(item.intent!!)
					true
				}
				item.itemId == R.id.image__choose_external__capture
					|| item.groupId == R.id.image__choose_external__capture_group -> {
					captureImage.launch(item.intent!!)
					true
				}
				item.itemId == R.id.image__choose_external__visual
					|| item.groupId == R.id.image__choose_external__visual_group -> {
					pickVisualImage.launch(item.intent!!)
					true
				}
				else -> {
					error("Unknown menu item: ${item}")
				}
			}
		}
	}

	fun show() {
		// Set dismiss listener right before showing to make sure it exists.
		// It might be cleared when an item is selected. See #setOnMenuItemClickListener.
		menu.setOnDismissListener { events.onCancelled() }
		populateGroups(menu.menu)
		menu.show()
	}

	private fun populateGroups(menu: Menu) {
		menu.add(R.id.image__choose_external__pick_group, R.id.image__choose_external__pick)
		menu.add(R.id.image__choose_external__get_group, R.id.image__choose_external__get)
		menu.add(R.id.image__choose_external__visual_group, R.id.image__choose_external__visual)
		menu.add(R.id.image__choose_external__capture_group, R.id.image__choose_external__capture)
		menu.asIterable().forEach {
			it.icon = it.icon?.fixIcon(it.isHeading, context.resources)
			if (it.isHeading) {
				it.title = TextTools.bold(it.title)
			}
		}
		menu.showCapture(CameraTools.canLaunchCameraIntent(context))
		menu.showPickVisual(isPhotoPickerAvailable(context))
	}

	interface Events {
		fun onCancelled()
		fun itemSelected()
		fun onGetContent(result: Uri)
		fun onPick(result: Uri)
		fun onPickVisualImage(result: Uri)
		fun onCapture(result: Uri)
	}
}

private class ExplicitAbleActivityResultLauncher<I>(
	private val context: Context,
	private val contract: ActivityResultContract<I, *>,
	private val input: I,
	private val launcher: ActivityResultLauncher<Intent>,
) {

	/**
	 * Create a generic intent for the [contract] and [input].
	 */
	fun createIntent(): Intent =
		contract.createIntent(context, input)

	/**
	 * Launch the generic intent generated by [createIntent],
	 * or launch a package-specific explicit intent (overridden component) from the same.
	 */
	fun launch(input: Intent) {
		launcher.launch(input)
	}

	companion object {
		// Cannot be a constructor because of generic parameters.
		operator fun <C, I, O> invoke(
			context: C,
			contract: ActivityResultContract<I, O>,
			input: I,
			callback: ActivityResultCallback<O>
		): ExplicitAbleActivityResultLauncher<I>
			where C : Context, C : ActivityResultCaller =
			ExplicitAbleActivityResultLauncher(
				context,
				contract,
				input,
				context.registerForActivityResult(
					ActivityResultContracts.StartActivityForResult()
				) {
					callback.onActivityResult(contract.parseResult(it.resultCode, it.data))
				}
			)
	}
}

private fun createMenu(activity: Activity, anchor: View): PopupMenu =
	PopupMenu(activity, anchor).apply {
		gravity = Gravity.TOP or Gravity.START
		setForceShowIcon(true) // See fixIcons().
		inflate(R.menu.image__choose_external)
	}

/**
 * Types of menu items:
 *  * Headings are not in a group, they're just above them.
 *  * Groups are represented in the [R.menu.image__choose_external] XML.
 *   They're populated by [add] and they have `groupId` set.
 *  * There's no third type yet, so being lazy with [Boolean].
 */
private val MenuItem.isHeading: Boolean
	get() = this.groupId == Menu.NONE

private fun Menu.add(@IdRes groupId: Int, @IdRes headerId: Int) {
	val header = this.findItem(headerId) ?: error("Menu item not found for id: ${headerId}")
	this.addIntentOptions(groupId, Menu.NONE, header.order + 1, null, null, header.intent, 0, null)
	when (this.asIterable().count { it.groupId == groupId }) {
		0 -> {
			header.isVisible = false
		}
		1 -> {
			header.isVisible = true
			//header.collapseFromSingletonGroup(this, groupId)
		}
		else -> {
			header.isVisible = true
		}
	}
}

@Suppress("unused") // Keep for now.
private fun MenuItem.collapseFromSingletonGroup(menu: Menu, groupId: Int) {
	val item = menu.asIterable().single { it.groupId == groupId }
	icon = item.icon
	menu.removeGroup(groupId)
}

private fun Menu.showCapture(canCapture: Boolean) {
	this.findItem(R.id.image__choose_external__capture).isVisible = canCapture
	this.setGroupVisible(R.id.image__choose_external__capture_group, canCapture)
}

private fun Menu.showPickVisual(hasPhotoPicker: Boolean) {
	this.findItem(R.id.image__choose_external__visual).isVisible = hasPhotoPicker
	this.setGroupVisible(R.id.image__choose_external__visual_group, hasPhotoPicker)
}

private fun isPhotoPickerAvailable(context: Context): Boolean =
	if (Build.VERSION_CODES.KITKAT <= Build.VERSION.SDK_INT)
		ActivityResultContracts.PickVisualMedia.isPhotoPickerAvailable(context)
	else
		false

/**
 * Ensure all drawables have the same size, otherwise they're misaligned.
 * Also add some indentation to resolved package listings, to give some hierarchy without use of submenus.
 *
 * [PopupMenu.show] via [androidx.appcompat.view.menu.MenuPopupHelper]
 * instantiates [androidx.appcompat.view.menu.StandardMenuPopup]
 * which inflates [androidx.appcompat.R.layout.abc_popup_menu_item_layout]
 * using [androidx.appcompat.view.menu.MenuAdapter].
 * Items in the adapter will be [androidx.appcompat.view.menu.ListMenuItemView]
 * and their icons are added in [androidx.appcompat.view.menu.ListMenuItemView.insertIconView]
 * from [androidx.appcompat.R.layout.abc_list_menu_item_icon].
 */
private fun Drawable.fixIcon(heading: Boolean, resources: Resources): Drawable {
	val indent = if (heading) 0 else
		resources.getDimensionPixelSize(R.dimen.image__choose_external__menu_icon_indent)
	val size = resources.getDimensionPixelSize(
		if (heading)
			R.dimen.image__choose_external__menu_icon_header_size
		else 
			R.dimen.image__choose_external__menu_icon_intent_size
	)
	return object : DrawableWrapperCompat(this) {
		// Adding indent reserves extra width, because ImageView will take this size.
		override fun getIntrinsicWidth(): Int = size + indent
		override fun getIntrinsicHeight(): Int = size
		override fun setBounds(left: Int, top: Int, right: Int, bottom: Int) {
			// Shift left by `indent`, so the icon gets back to it's `size` width, and the left side of it will be empty.
			super.setBounds(left + indent, top, right, bottom)
		}
	}
}
